### PR TITLE
[runtime_env] Fix the some bugs related with runtime_env

### LINKED ISSUE #15254: [runtime_env] Bad error message for non-existent working_dir
Passing a working_dir that doesn't exist gives me the following:
```python
~/code/ray/python/ray/_private/runtime_env.py in get_project_package_name(working_dir, py_modules)
    185     if working_dir:
    186         assert isinstance(working_dir, str)
--> 187         assert Path(working_dir).exists()
    188         hash_val = _xor_bytes(hash_val, _hash_modules(Path(working_dir)))
    189     for py_module in py_modules or []:

AssertionError: 
```

### LINKED ISSUE #15255: [runtime_env] Passing an empty working directory errors
```python
# tmp2 is an empty directory.
ray.init(address="auto", job_config=ray.job_config.JobConfig(runtime_env={"working_dir": "tmp2"}))
```

```
~/code/ray/python/ray/_private/runtime_env.py in rewrite_working_dir_uri(job_config)
    311         pkg_name = get_project_package_name(working_dir, py_modules)
    312         job_config.runtime_env[
--> 313             "working_dir_uri"] = Protocol.GCS.value + "://" + pkg_name
    314 
    315 

TypeError: can only concatenate str (not "NoneType") to str
```

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
There are three issues reported related to runtime_env:
- empty dir doesn't work
- bad error message
- util without job config failed to find module

## Related issue number
Closes #15254 
Closes #15255 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/runtime_env.py [hunk 1]
```diff
@@ -213,13 +213,17 @@ def get_project_package_name(working_dir: str, py_modules: List[str],
     hash_val = None
     excludes = {Path(p).absolute() for p in excludes}
     if working_dir:
-        assert isinstance(working_dir, str)
-        assert Path(working_dir).exists()
+        assert isinstance(working_dir, str), f"`working_dir` is not a string."
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:216]: Please don't use `assert` statements for user-facing errors. Best practice here is to use `TypeError` for a type check and `ValueError` for a value check (e.g., an int must be >= 0 or the directory must exist)

File: python/ray/_private/runtime_env.py [hunk 2]
```diff
@@ -213,13 +213,17 @@ def get_project_package_name(working_dir: str, py_modules: List[str],
...
         working_dir = Path(working_dir).absolute()
+        if not working_dir.exists() or not working_dir.is_dir():
+            raise IOError(f"Invalid working dir {working_dir}.")
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:219]:
```suggestion
            raise ValueError(f"working_dir {working_dir} must be an existing directory.")
```

File: python/ray/_private/runtime_env.py [hunk 3]
```diff
@@ -213,13 +213,17 @@ def get_project_package_name(working_dir: str, py_modules: List[str],
...
         hash_val = _xor_bytes(
             hash_val, _hash_modules(working_dir, working_dir, excludes))
     for py_module in py_modules or []:
+        assert isinstance(py_module, str), f"`py_module` is not a string."
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:223]: TypeError

File: python/ray/_private/runtime_env.py [hunk 4]
```diff
@@ -213,13 +213,17 @@ def get_project_package_name(working_dir: str, py_modules: List[str],
...
         module_dir = Path(py_module).absolute()
+        if not module_dir.exists() or not module_dir.is_dir():
+            raise IOError(f"Invalid py_module {py_module}.")
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:226]: ValueError. Please also make the error more actionable:
```suggestion
            raise ValueError(f"py_module {py_module} must be a path to an existing directory.")
            raise IOError(f"Invalid py_module {py_module}.")
```

File: python/ray/_private/runtime_env.py [hunk 5]
```diff
@@ -213,13 +213,17 @@ def get_project_package_name(working_dir: str, py_modules: List[str],
     hash_val = None
     excludes = {Path(p).absolute() for p in excludes}
     if working_dir:
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:215]: Add some basic tests for these checks to avoid regressions please.

Comment (edoakes) [reply python/ray/_private/runtime_env.py:215]: You can use `with pytest.raises(ValueError): ...`

**Review by edoakes — CHANGES_REQUESTED** [review]

**Comment by ericl** [comment]
//python/ray/tests:test_runtime_env                                      FAILED in 3 out of 3 in 393.4s
  Stats over 3 runs: max = 393.4s, min = 375.7s, avg = 382.2s, dev = 8.0s
  /private/var/tmp/_bazel_travis/7c557718f3877739c657a427203800b1/execroot/com_github_ray_project_ray/bazel-out/darwin-opt/testlogs/python/ray/tests/test_runtime_env/test.log
  /private/var/tmp/_bazel_travis/7c557718f3877739c657a427203800b1/execroot/com_github_ray_project_ray/bazel-out/darwin-opt/testlogs/python/ray/tests/test_runtime_env/test_attempts/attempt_1.log
  /private/var/tmp/_bazel_travis/7c557718f3877739c657a427203800b1/execroot/com_github_ray_project_ray/bazel-out/darwin-opt/testlogs/python/ray/tests/test_runtime_env/test_attempts/attempt_2.log
Executed 47 out of 47 tests: 46 tests pass and 1 fails locally.

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
Trusting the tests-ok flag from @iycheng

**Pull request merged**